### PR TITLE
fix(microfloat): make e4m3 conversion bit-correct for OCP OFP8

### LINKED ISSUE #1302: microfloat e4m3 is not bit-correct for OCP E4M3 (overflow and ±inf saturate instead of NaN); non-saturating variant maps inf to zero
`microfloat<8,4,false,true,true>` (aliased `e4m3`) is very close to the OCP
8-bit E4M3 format, but is **not bit-correct** for conversions from a wider type.
The encoding itself is already exact — the gap is entirely in the
conversion-from-wider policy, in three specific places.

Target is the [OCP 8-bit Floating Point Specification](https://www.opencompute.org/documents/ocp-8-bit-floating-point-specification-ofp8-revision-1-0-2023-12-01-pdf-1)
E4M3 (commonly `e4m3fn` — *finite*, no infinities, NaN only at `S.1111.111`).
`ml_dtypes.float8_e4m3fn` is used below purely as a cross-check because it
implements that spec, as do JAX and PyTorch; the standard is the target, not any
particular library.

## Already correct — worth stating, since it scopes the fix

Decoding is **bit-exact across the entire encoding space**: all **256 of 256**
patterns decode to the same value as the reference, including subnormals, zero,
`maxpos = 448` at `0x7e`, and NaN at `0x7f`/`0xff`. Round-to-nearest on finite
input is correct too, including the tie at the top of the range: `464.0` (the
exact midpoint above 448) rounds to `448` in both, as round-to-nearest-even
requires.

So no re-encoding is needed. Only the out-of-range and non-finite conversion
paths need changing.

## The three divergences

| input | `microfloat` e4m3 | OCP e4m3fn | |
|-------|-------------------|------------|---|
| `> 464.0` (e.g. `500`, `1e30`) | `0x7e` (saturates to 448) | `0x7f` (NaN) | ✗ |
| `±inf` | `0x7e` / `0xfe` (±448) | `0x7f` / `0xff` (NaN, sign kept) | ✗ |
| `-NaN` | `0x7f` (+NaN) | `0xff` (sign preserved) | ✗ |
| `+NaN` | `0x7f` | `0x7f` | ✓ |
| `464.0`, `449`, `448` | `0x7e` | `0x7e` | ✓ |

1. **Finite overflow saturates instead of becoming NaN.** `isSaturating = true` in
   the alias clamps anything above the range to ±448. E4M3's defining property is
   that it has no infinity *and* signals overflow as NaN.
2. **`±inf` saturates to ±448.** It should become NaN with the sign preserved.
3. **`-NaN` loses its sign**, landing on `0x7f` instead of `0xff`. Minor, but it
   means a NaN does not round-trip bit-for-bit through the type.

## Separate bug: the non-saturating variant maps ±inf to zero

While checking whether an existing configuration already had the right behavior,
`microfloat<8,4,false,true,false>` (same but `isSaturating = false`) turned out
to be worse:

```
config                                     448   500   1e30  -1e30  +inf  -inf
<8,4,noInf,NaN,saturating>   (e4m3 alias)  0x7e  0x7e  0x7e  0xfe   0x7e  0xfe
<8,4,noInf,NaN,non-saturating>             0x7e  0x7e  0x7e  0xfe   0x00  0x00   <-- inf -> ZERO
ml_dtypes.float8_e4m3fn (reference)        0x7e  0x7f  0x7f  0xff   0x7f  0xff
```

`+inf` and `-inf` convert to **`0x00`, i.e. +0**. That is wrong under any policy
— saturation would give maxpos, NaN-on-overflow would give NaN, and zero is
neither. It also silently turns an overflow into the most benign possible value.
Worth fixing independently of the E4M3 question, since it affects every
non-saturating `microfloat` config with `hasInf = false`.

(The `hasInf = true` configs are a different format — `maxpos` drops to 256 at
`0x78` — so they are not a route to E4M3.)

## Suggested direction

Saturation-on-overflow is a legitimate mode and is genuinely wanted in some
inference pipelines, so the fix is probably not to change `e4m3`'s behavior in
place but to make the standard-conforming behavior reachable and correctly named:

- an `e4m3fn` alias whose conversion produces NaN on finite overflow and on
  `±inf`, with the sign preserved — matching the OCP spec bit-for-bit
- keep the saturating behavior available under a name that says so
  (`e4m3_saturating` or similar), rather than having `e4m3` mean the saturating
  variant while the ecosystem reads `e4m3` as the OCP format
- fix the `hasInf = false, isSaturating = false` inf → 0 case regardless

A conformance test over all 256 decodings plus the conversion cases in the table
above would pin it; the decode half already passes.

## Why this matters downstream

[universal_dtypes](https://github.com/stillwater-sc/universal_dtypes) exposes
Universal's number systems as NumPy dtypes. E4M3 is one of the two FP8 formats in
active use for ML inference, and is the one we cannot currently ship, because
shipping something named `e4m3` that saturates where the standard says NaN would
be worse than shipping nothing. `e5m2` we do ship. Once this is bit-correct we
can add the dtype and cross-check it the same way.

## Repro

```cpp
#include <universal/number/microfloat/microfloat.hpp>
using namespace sw::universal;

e4m3 a(500.0);            // -> 0x7e (448); OCP: 0x7f (NaN)
e4m3 b(1.0/0.0);          // -> 0x7e (448); OCP: 0x7f (NaN)
e4m3 c(-std::nan(""));    // -> 0x7f (+NaN); OCP: 0xff

microfloat<8,4,false,true,false> d(1.0/0.0);   // -> 0x00 (+0)  <-- separate bug
```

All figures above were produced by compiling against the current `main`
(`v4.7.9` + 6) and diffing against `ml_dtypes.float8_e4m3fn` over the full
encoding space.

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

`microfloat`'s E4M3 *encoding* was already exact -- all 256 patterns decode to the specification's value -- but conversion from a wider type diverged from the [OCP 8-bit Floating Point Specification](https://www.opencompute.org/documents/ocp-8-bit-floating-point-specification-ofp8-revision-1-0-2023-12-01-pdf-1) in the three places #1302 lists. A fourth turned up in `e5m2` while fixing them.

| input | was | OCP / IEEE |
|-------|-----|------------|
| `500.0` | `0x7e` (448) | `0x7f` (NaN) |
| `+inf` / `-inf` | `0x7e` / `0xfe` | `0x7f` / `0xff` |
| `-NaN` | `0x7f` | `0xff` |
| `e5m2` `58000.0` | `0x7c` (inf) | `0x7b` (57344) |
| `<8,4,false,true,false>` `+-inf` | `0x00` (+0) | -- plain bug |

## Root cause

All of them come from one block. `from_float()` pre-clamped against maxpos *before* rounding, short-circuiting the post-rounding overflow check underneath it. That check was already correct -- it rounds to nearest-even against the format's exponent range, so `464.0` (the tie between 448 and the 480 the NaN slot took) rounds down to maxpos and `465.0` overflows, exactly as the spec requires. The fix is to delete the pre-clamp and let it decide.

What to *do* about an overflow was spelled out three times and the copies did not agree. It is now one `setoverflow()` policy:

| | |
|---|---|
| `isSaturating` | maxpos / maxneg |
| `hasInf` | `+-inf` |
| `hasNaN`, no inf | NaN, sign preserved -- the OCP `fn` rule |
| none of the three | maxpos / maxneg, the only thing the format can say |

The last case used to be zero, which is why every non-saturating configuration without infinity turned `+-inf` into `+0`.

## The two e4m3 policies

Identical encoding, different conversion:

- **`e4m3` (= `e4m3fn`)** -- NaN on overflow, per OCP. What `ml_dtypes.float8_e4m3fn`, JAX and PyTorch implement, and what [universal_dtypes](https://github.com/stillwater-sc/universal_dtypes) needs in order to ship an `e4m3` NumPy dtype.
- **`e4m3_saturating`** -- clamps. MX and NVFP4 scale the block maximum into a range whose top lies above 448, so the largest element of a block routinely lands past maxpos and has to clip rather than poison the block with a NaN.

`mxfp8` and the NVFP4 block scale are pinned to `e4m3_saturating` and are **unchanged bit-for-bit**; the bare `e4m3` name now means the standard format, as #1302 recommends.

## Changes

- `microfloat_impl.hpp` -- `setoverflow()` policy; sign extracted before the NaN branch; `setnan(type, sign)`; pre-clamp deleted; the three overflow arms and `setinf`-without-infinity routed through the policy
- `microfloat_fwd.hpp` -- `e4m3fn`, `e4m3_saturating`, `e4m3 = e4m3fn`, `is_e4m3_encoding`
- `numeric_limits.hpp` -- `infinity()` returns `T()` when `has_infinity` is false, per the std convention, now that `setinf()` answers with the overflow policy rather than zero
- `mxfloat_fwd.hpp` / `nvblock_fwd.hpp` / both `manipulators.hpp` -- `mxfp8` and the NVFP4 scale pinned to the saturating policy; type tags know both
- `static/block/microfloat/conversion/ocp_e4m3_conformance.cpp` (new) -- all 256 decodings against a table generated from the specification's field layout, plus conversion against a brute-force nearest-with-ties-to-even search over that table. Neither oracle consults `microfloat` to decide what the answer is.
- `static/block/microfloat/api/constexpr.cpp` -- the overflow policy at compile time (the constexpr path extracts IEEE 754 fields by hand instead of calling `frexp`/`signbit`, so it has to reach the same verdict independently); the local aliases that shadowed the library's are gone
- `docs/number-systems/microfloat.md` -- the two policies, and the e4m3 range corrected from 240 to 448

## Oracle mutation test

Run against the pre-fix implementation, the new suite reports the decode half green and every divergence red -- which is the claim #1302 makes:

```
e4m3 decode, all 256                PASS
e4m3 conversion, OCP cases          FAIL 8 failed test cases
e4m3 conversion, vs oracle          FAIL 2 failed test cases
e4m3_saturating policy              PASS
e5m2 overflow boundary              FAIL 3 failed test cases
e4m3 infinite source                FAIL 3 failed test cases
no inf, no nan, no saturation       FAIL 3 failed test cases
e4m3 dense sweep                    FAIL 458752 failed test cases
```

## Test Results

All 23 microfloat + mxfloat + nvblock tests, both compilers, plus the new suite at all four regression levels standalone (the dense sweep at level 4 walks the whole finite range at 2^-12 against the oracle).

| Target | gcc build | gcc test | clang build | clang test |
|--------|-----------|----------|-------------|------------|
| microfloat (10 targets) | OK | PASS | OK | PASS |
| mxfloat (8 targets) | OK | PASS | OK | PASS |
| nvblock (5 targets) | OK | PASS | OK | PASS |
| ocp_e4m3_conformance, levels 1-4 | OK | PASS | OK | PASS |

`tools/scripts/check-ascii.sh`: clean.

## Test plan
- [ ] Fast CI passes (gcc + clang CI_LITE)
- [ ] Promote to ready when satisfied

Resolves #1302

Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added distinct non-saturating `e4m3` and saturating `e4m3_saturating` formats.
  * `e4m3` now produces NaN on overflow, while `e4m3_saturating` clamps to the maximum finite value.
  * Updated MXFP8 and NVFP4 defaults to use saturating E4M3 behavior.
  * Improved handling of signed NaNs, infinities, and overflow across supported microfloat formats.

* **Documentation**
  * Clarified overflow behavior and usage for OCP, MX, NVIDIA, JAX, and PyTorch formats.

* **Tests**
  * Added comprehensive conformance and conversion coverage for rounding, overflow, NaNs, infinities, and boundary values.

### CODE REVIEW

File: static/block/microfloat/conversion/ocp_e4m3_conformance.cpp [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+	// a dense sweep of the whole finite range against the oracle, at a step
+	// well below the smallest subnormal
+	{
+		int nrOfFailedTests = 0;
+		for (float x = 0.0f; x <= 520.0f; x += 0.000244140625f) {   // 2^-12
+			e4m3 a(x), b(-x);
+			nrOfFailedTests += check(reportTestCases, "e4m3 sweep", x, a.bits(), ocp_encode(x));
+			nrOfFailedTests += check(reportTestCases, "e4m3 sweep", -x, b.bits(), ocp_encode(-x));
+		}
+		nrOfFailedTestCases += ReportTestResult(nrOfFailedTests, test_tag, "e4m3 dense sweep");
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply static/block/microfloat/conversion/ocp_e4m3_conformance.cpp:392]: Not taking this one. The iteration count is right, the cost is not: measured at `-O2`, the whole suite with all four levels enabled runs in **0.256 s**, and the CMake-built binary that CI actually runs takes **0.001 s**, because CMake always passes `REGRESSION_LEVEL_OVERRIDE` and enables only level 1.

```
$ time ./ocp_e4m3_conformance                        # hand-compiled, all four levels
OCP OFP8 E4M3 conformance: PASS
real    0m0.262s

$ time ./build_mf/.../microfloat_ocp_e4m3_conformance   # CMake, level 1
OCP OFP8 E4M3 conformance: PASS
real    0m0.001s
```

The `#define REGRESSION_LEVEL_4 1` fallback is the repo's standard test skeleton -- every regression `.cpp` in the tree carries it verbatim, and it exists precisely so a hand-compiled binary runs everything while CI runs the level it asked for. Setting it to 0 here would make this the one file that silently skips its own level-4 section when a developer compiles it directly, which is the opposite of what the skeleton is for. Level 4 is documented in this repo as "exhaustive stress testing"; a quarter of a second sits well inside that budget.

**Comment by Ravenwater** [comment]
@coderabbitai review
